### PR TITLE
Add OpenStack provider support with configuration and templates

### PR DESCRIPTION
**Pull request by bnallapeta**
Resolves https://github.com/Mirantis/hmc/issues/594.

OpenStack Provider Integration:
- Introduces OpenStack as a supported cloud provider in HMC.
- Adds configurations and provider templates specific to OpenStack environments.

Local Run:

```
ka get machines
NAME                           CLUSTER         NODENAME                       PROVIDERID                                          PHASE     AGE     VERSION
openstack-dev-cp-0             openstack-dev   openstack-dev-cp-0             openstack:///b52ccc71-566b-4487-a0b8-64b2c08f0e5b   Running   8m29s   v1.30.5+k0s.0
openstack-dev-cp-1             openstack-dev   openstack-dev-cp-1             openstack:///4022efcb-da46-4a99-80f3-2eaf3681238e   Running   8m29s   v1.30.5+k0s.0
openstack-dev-cp-2             openstack-dev   openstack-dev-cp-2             openstack:///87ae87e0-2ac5-4f06-b7ba-970d6dbcab3f   Running   8m29s   v1.30.5+k0s.0
openstack-dev-md-p27j6-6wfn4   openstack-dev   openstack-dev-md-p27j6-6wfn4   openstack:///5c2c51c8-f2c0-48d3-9c8f-ed28db53112a   Running   8m30s   
openstack-dev-md-p27j6-w6qvx   openstack-dev   openstack-dev-md-p27j6-w6qvx   openstack:///2df39a6d-2507-452f-9a0a-fb3e6a5d2052   Running   8m30s   
```

```
clusterctl describe cluster openstack-dev -n hmc-system
NAME                                                      READY  SEVERITY  REASON  SINCE  MESSAGE                                                        
Cluster/openstack-dev                                     True                     5m26s                                                                  
├─ClusterInfrastructure - OpenStackCluster/openstack-dev                                                                                                  
├─ControlPlane - K0sControlPlane/openstack-dev-cp                                                                                                         
│ └─3 Machines...                                         True                     7m32s  See openstack-dev-cp-0, openstack-dev-cp-1, ...                 
└─Workers                                                                                                                                                 
  └─MachineDeployment/openstack-dev-md                    True                     110s                                                                   
    └─2 Machines...                                       True                     3m43s  See openstack-dev-md-p27j6-6wfn4, openstack-dev-md-p27j6-w6qvx  
```

```
k -n hmc-system get managedclusters.hmc.mirantis.com 
NAME            READY   STATUS
openstack-dev   True    ManagedCluster is ready
```

### CODE REVIEW

File: templates/provider/cluster-api-provider-openstack/templates/secret.yaml
Comment (eromanova) [inline templates/provider/cluster-api-provider-openstack/templates/secret.yaml:8]: Could you please add an empty `config` to the `values.yaml` file and add schema for it?

File: templates/provider/cluster-api-provider-openstack/values.schema.json
Comment (eromanova) [inline templates/provider/cluster-api-provider-openstack/values.schema.json:15]: I'm not sure the `namespace` parameter should be required. The `Release.Namespace` will be used by default if unset.

**Review by eromanova — CHANGES_REQUESTED** [review]

File: templates/provider/cluster-api-provider-openstack/values.schema.json
Comment (bnallapeta) [reply templates/provider/cluster-api-provider-openstack/values.schema.json]: Yes, it does pick up the Release.Namespace by default. Removed this.

File: templates/provider/cluster-api-provider-openstack/templates/secret.yaml
Comment (bnallapeta) [reply templates/provider/cluster-api-provider-openstack/templates/secret.yaml:8]: Done.

File: cmd/main.go
Comment (a13x5) [inline cmd/main.go:31]: Why it's needed?

File: cmd/main.go
Comment (a13x5) [inline cmd/main.go:59]: I don't see any related changes to the controller which are using openstack types, so it looks completely out of scope of this PR.
Remove.

File: templates/provider/cluster-api-provider-openstack/values.yaml
Comment (a13x5) [inline templates/provider/cluster-api-provider-openstack/values.yaml:2]: Why do we need to create an empty config?

**Review by a13x5 — CHANGES_REQUESTED** [review]
Weird imports should be removed (or moved to another PR).
Minor clarifications are needed.

File: templates/provider/cluster-api-provider-openstack/values.yaml
Comment (bnallapeta) [reply templates/provider/cluster-api-provider-openstack/values.yaml:2]: we are using config in [secret.yaml](https://github.com/bnallapeta/hmc/blob/openstack-add-provider/templates/provider/cluster-api-provider-openstack/templates/secret.yaml#L8). If this is not defined, helm template will throw an error. It also lets the users know what fields are available in values.yaml

File: cmd/main.go
Comment (bnallapeta) [reply cmd/main.go:59]: Same as above.

File: templates/provider/cluster-api-provider-openstack/values.yaml
Comment (a13x5) [reply templates/provider/cluster-api-provider-openstack/values.yaml:2]: I mean you have empty `.Values.config` by default and thus secret will be created with and empty data (which is weird). So if controller will work fine without the empty secret I suggest setting `configSecret.create: false` and skip the creation of the secret

File: go.mod
Comment (a13x5) [inline go.mod:1]: If imports aren't needed, dependencies in the `go.mod/go.sum` aren't needed as well. Remove.

**Review by a13x5 — CHANGES_REQUESTED** [review]

File: templates/provider/cluster-api-provider-openstack/values.yaml
Comment (bnallapeta) [reply templates/provider/cluster-api-provider-openstack/values.yaml:2]: Yes, I did test it with the current way which is an empty config. But I will test this again to make sure it works and then make it false.

File: go.mod
Comment (bnallapeta) [reply go.mod:1]: Done.

Comment (a13x5) [reply go.mod:1]: Still see them in the diff.

Comment (bnallapeta) [reply go.mod:1]: Indirect dependencies. Result of `go mod tidy` and latest changes from main.

File: templates/provider/cluster-api-provider-openstack/values.yaml
Comment (bnallapeta) [reply templates/provider/cluster-api-provider-openstack/values.yaml:2]: tested this. it works with `create:false`

File: config/dev/openstack-managedcluster.yaml
Comment (eromanova) [inline config/dev/openstack-managedcluster.yaml:12]: It does not match the configuration defined in the `openstack-standalone-cp` helm chart (here we have `region` but in the cluster template helm chart we have `clusterIdentity.region` https://github.com/Mirantis/hmc/pull/696/files). 

Anyway, I suggest moving this OpenStack ManagedCluster testing YAML to a separate PR. Currently, we lack both an OpenStack standalone cluster template and the required credentials handling. It isn’t critical, but I prefer patches to focus only on related changes for better clarity and organization.

**Review by eromanova — COMMENTED** [review]
One comment from my side. The rest looks good.

File: config/dev/openstack-managedcluster.yaml
Comment (bnallapeta) [reply config/dev/openstack-managedcluster.yaml:12]: so should I move this file to the cluster template PR?

**Review by eromanova — APPROVED** [review]

**Comment by bnallapeta** [comment]
@a13x5 @eromanova Would appreciate if you can take a look at the PR and merge it soon.

**Comment by eromanova** [comment]
> @a13x5 @eromanova Would appreciate if you can take a look at the PR and merge it soon.

LGTM. Waiting for Alexey's approval since he also requested changes :)

**Comment by a13x5** [comment]
@bnallapeta @eromanova Let's merge it after airgap fixes will be fully landed. 
Meanwhile I believe the `cm-ag-resources.yaml` file from #687  could be incorporated (no changes in file required) in this PR, since files will be collected, but not used.

**Review by a13x5 — APPROVED** [review]
Approved
As discussed additional changes/fixes could be added later

File: templates/provider/hmc-templates/files/templates/cluster-api-provider-openstack.yaml
Comment (eromanova) [inline templates/provider/hmc-templates/files/templates/cluster-api-provider-openstack.yaml:7]: Could you please re-run `make templates-generate`? The Template struct has been changed and should be updated.

**Review by eromanova — CHANGES_REQUESTED** [review]
LGTM. Please, address the comment and we can proceed with merging.

File: templates/provider/hmc-templates/files/templates/cluster-api-provider-openstack.yaml
Comment (bnallapeta) [reply templates/provider/hmc-templates/files/templates/cluster-api-provider-openstack.yaml:7]: Done.

**Review by eromanova — APPROVED** [review]

**Pull request merged**
